File: Week2 Assignment/dynamic_array_operation.py
#Dynamic Array operation

arr = []

# Insert elements
arr.append(10)
arr.append(20)
arr.append(30)

print("After insertion:", arr)

# Delete an element
arr.remove(20)  # removes first occurrence of 20
print("After deletion:", arr)

# Access
print("Element at index 1:", arr[1])

# Resize happens automatically (append or pop)
arr.append(40)
arr.append(50)
print("After more insertions:", arr)

# Remove last element
arr.pop()
print("After popping:", arr)

File: Week2 Assignment/dynamic_array_operation.c
//Dynamic Array Operation

#include <stdio.h>
#include <stdlib.h>

int main() {
    int *arr;
    int capacity = 2;  // initial capacity
    int size = 0;      // current size
    int i;

    // Allocate initial memory
    arr = (int *)malloc(capacity * sizeof(int));
    if (arr == NULL) {
        printf("Memory allocation failed!\n");
        return 1;
    }

    // Insert elements with dynamic resizing
    for (i = 0; i < 5; i++) {
        if (size == capacity) {
            capacity *= 2;
            arr = (int *)realloc(arr, capacity * sizeof(int));
            if (arr == NULL) {
                printf("Memory reallocation failed!\n");
                return 1;
            }
        }
        arr[size++] = (i + 1) * 10;
    }

    // Print elements
    printf("Array elements: ");
    for (i = 0; i < size; i++) {
        printf("%d ", arr[i]);
    }

    // Delete last element
    size--;
    printf("\nAfter deletion: ");
    for (i = 0; i < size; i++) {
        printf("%d ", arr[i]);
    }

    // Access an element
    printf("\nElement at index 2: %d\n", arr[2]);

    // Free memory
    free(arr);
    return 0;
}

